### ISSUE TITLE
Undefined AWS::ApiGateway

### ISSUE THREAD

**Issue by AlGreed** [dougmoscrop/serverless-plugin-split-stacks#37]
Hi,

Since last release there has been the following error:
 
```
  Type Error ---------------------------------------------
 
  Cannot set property 'AWS::ApiGateway::Resource' of undefined
```

In stack-map.js we turn off creation of nested stack for ApiGateway:
````
const stacksMap = require('serverless-plugin-split-stacks').stacksMap;

//ignore Api-Gateway for nesting stacks
stacksMap['AWS::ApiGateway::Resource'] = undefined;
stacksMap['AWS::ApiGateway::RestApi'] = undefined;
````

**Comment by dougmoscrop** [comment]
Sorry about that, can you try with 1.5.1?

**Comment by Rapol** [comment]
Our deploys stopped working today after been updated to 1.5.1. I pinned the library back to 1.5.0 and it is working again.

Error:
```
Serverless: Invoke aws:package:finalize
 
  Type Error ---------------------------------------------
 
  Cannot read property 'AWS::S3::Bucket' of undefined
 
     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
 
  Stack Trace --------------------------------------------
 
TypeError: Cannot read property 'AWS::S3::Bucket' of undefined
    at ServerlessPluginSplitStacks.resolveMigration (/home/ubuntu/workspace/stacks-map.js:23:26)
    at Custom.migration (/home/ubuntu/workspace/node_modules/serverless-plugin-split-stacks/lib/migration-strategy/custom.js:11:55)
    at _.each (/home/ubuntu/workspace/node_modules/serverless-plugin-split-stacks/lib/migrate-new-resources.js:20:30)
    at /home/ubuntu/workspace/node_modules/lodash/lodash.js:4925:15
    at baseForOwn (/home/ubuntu/workspace/node_modules/lodash/lodash.js:3010:24)
    at /home/ubuntu/workspace/node_modules/lodash/lodash.js:4894:18
    at Function.forEach (/home/ubuntu/workspace/node_modules/lodash/lodash.js:9342:14)
    at ServerlessPluginSplitStacks.migrateResources [as migrateNewResources] (/home/ubuntu/workspace/node_modules/serverless-plugin-split-stacks/lib/migrate-new-resources.js:14:5)
    at Promise.resolve.then.then (/home/ubuntu/workspace/node_modules/serverless-plugin-split-stacks/split-stacks.js:61:24)
```

stack-maps.js
```
const ServerlessPluginSplitStacks = require('serverless-plugin-split-stacks');

ServerlessPluginSplitStacks.resolveMigration = function (resource, logicalId, serverless) {

    if(resource.Type === "AWS::Lambda::Permission") return null;
    if(resource.Type === "AWS::Lambda::Version") return null;
    if(resource.Type === "AWS::CloudWatch::Alarm") return null;
    if(resource.Type === "AWS::Logs::MetricFilter") return null;
    if(resource.Type === "AWS::Logs::SubscriptionFilter") return null;
    if(resource.Type === "AWS::SNS::Subscription") return null;
    if(resource.Type === "AWS::SNS::TopicPolicy") return null;
    if(resource.Type === "AWS::S3::BucketPolicy") return null;
    if(resource.Type === "AWS::SQS::QueuePolicy") return null;

    if (logicalId.startsWith("ApiGatewayMethodAccounts")) return {destination: 'AccountsMethods'};
    if (logicalId.startsWith("ApiGatewayMethodThings")) return {destination: 'ThingsMethods'};
    if (resource.Type == 'AWS::ApiGateway::Method') return {destination: 'GeneralMethods'};
    if (resource.Type == 'AWS::ApiGateway::Model') return {destination: 'Models'};

    // Fallback to default:
    return this.stacksMap[resource.Type];
};
```

**Comment by dougmoscrop** [comment]
Sorry about that. Working on a fix.

**Comment by Rapol** [comment]
Pardon my ignorance, I just noticed that the library was updated with more default ways to split the stack. Since Im still using the stack-maps.js I set both perType and perLambda to false. I looked at the error and found that the problem is that the stackMaps was never initialized. 

On the last commit, a change to lib/migrate-new-resources.js(line 11) was made that removed this `this.stacksMap = {};`

I modified my local and added this and was able to deploy. Didn't ran any tests and not sure how this affects the new stuff.

**Comment by dougmoscrop** [comment]
Yeah, the problem is how it gets read from too. I should have a fix out
soon.

**Comment by dougmoscrop** [comment]
Did 1.5.2 work for you?

**Comment by medikoo** [comment]
@dougmoscrop Our setup just broken with v1.5.2.

Firstly it's because the things we discussed before, that options handling changed without bumping the major.

Now suddenly for new deploymets we get just few nested stacks, and 241 resources at root, which obviously crashes.

It would be really way better if this package follows semver

**Comment by dougmoscrop** [comment]
What's the last version that worked for you?

**Comment by Rapol** [comment]
Sorry for the delay, 1.5.2 works for me. Thanks for the fast response. Awesome plugin btw, cant believe is not that popular.

**Comment by dougmoscrop** [comment]
I tried to make it semver fit but these globals are not a good Patten.

I'm gonna deprecate all of 1.5 and release 2.0 with the changes I wated to
make anyway

**Comment by medikoo** [comment]
> but these globals are not a good Patten.

what globals? I don't understand these sentence.

> I'm gonna deprecate all of 1.5 and release 2.0 with the changes I wated to make anyway

It'll be great if proper support for customization of `Plugin.resolveMigration` is brought back (currently it doesn't work, so to make code clean either it should be removed or fixed)

**Comment by dougmoscrop** [comment]
Can you confirm 1.5.2 does not work? I thought I got it back to behavior

Globals was the wrong term. I meant statics.

**Comment by medikoo** [comment]
> Can you confirm 1.5.2 does not work?

It does not. I looked into it closer, and it was actually broken by this commit -> https://github.com/dougmoscrop/serverless-plugin-split-stacks/commit/5d3352cf11cfbc90e8ab722fcac161494cf1fd3b#diff-bf13709aa5d8df739da6faa9e893e786R3

It introduced cyclic dependency, and as`Plugin` is required before module at `../..` is exported, we get the plain object assigned to `Plugin`

So either that should stay as it was before change, or `require('../../')` should be done within a body of `migration` method (then it'll return exported plugin class)

**Comment by medikoo** [comment]
> Try 1.5.3? 

Nothing changed, as issue I [described](https://github.com/dougmoscrop/serverless-plugin-split-stacks/issues/37#issuecomment-390973984) was still not addressed.

Add `console.log(Plugin)` at https://github.com/dougmoscrop/serverless-plugin-split-stacks/blob/95ead539b7e85734adf16f1c4a810d40a7fd4108/lib/migration-strategy/custom.js#L6 you'll see it's not really a _Plugin_ as it's assumed

**Comment by dougmoscrop** [comment]
Yeah, sorry it took so long to fix, @medikoo - 1.5.4 removes the circular dep

**Comment by medikoo** [comment]
@dougmoscrop thank you! It works now
